common/logger/src/lib.rs

```diff
@@ -1,74 +1,426 @@
+// It have few unsafe lines, but does not require extra dependency
+static NOP: NopStructLog = NopStructLog {};
+static mut STRUCT_LOGGER: &'static dyn StructLogSink = &NOP;
+static STRUCT_LOGGER_STATE: AtomicUsize = AtomicUsize::new(UNINITIALIZED);
+const UNINITIALIZED: usize = 0;
+
+const INITIALIZING: usize = 1;
+const INITIALIZED: usize = 2;
+
+#[derive(Default, Serialize)]
+pub struct StructuredLogEntry {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    text: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pattern: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    name: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    module: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    location: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    git_rev: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    timestamp: Option<String>,
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    data: HashMap<&'static str, Value>,
+}
+
+#[must_use = "use StructuredLogEntry::send to send structured log"]
+impl StructuredLogEntry {
+    pub fn new_unnamed() -> Self {
+        Self::default()
+    }
+
+    pub fn new_named(name: &'static str) -> Self {
+        let mut ret = Self::default();
+        ret.name = Some(name);
+        ret
+    }
+
+    pub fn text(&mut self, text: String) -> &mut Self {
+        self.text = Some(text);
+        self
+    }
+
+    pub fn pattern(&mut self, pattern: &'static str) -> &mut Self {
+        self.pattern = Some(pattern);
+        self
+    }
+
+    pub fn module(&mut self, module: &'static str) -> &mut Self {
+        self.module = Some(module);
+        self
+    }
+
+    pub fn location(&mut self, location: &'static str) -> &mut Self {
+        self.location = Some(location);
+        self
+    }
+
+    pub fn git_rev(&mut self, git_rev: Option<&'static str>) -> &mut Self {
+        self.git_rev = git_rev;
+        self
+    }
+
+    pub fn timestamp(&mut self, timestamp: String) -> &mut Self {
+        self.timestamp = Some(timestamp);
+        self
+    }
+
+    pub fn json_data(&mut self, key: &'static str, value: Value) -> &mut Self {
+        self.data.insert(key, value);
+        self
+    }
+
+    pub fn data<D: Serialize>(&mut self, key: &'static str, value: D) -> &mut Self {
+        self.data.insert(
+            key,
+            serde_json::to_value(value).expect("Failed to serialize StructuredLogEntry key"),
+        );
+        self
+    }
+
+    // Use send_struct_log! macro instead of this method to populate extra meta information such as git rev and module name
+    pub fn send(self) {
+        struct_logger().send(self);
+    }
+}
+
+// This is exact copy of similar function in log crate
+pub fn set_struct_logger(logger: &'static dyn StructLogSink) -> Result<(), ()> {
```

Comment:
This makes sense for MVP, but I would think at some point we might want to have a common entry point to our thread that takes in a log and it then routes the structured logging to the structured logging handler and the unstructured to the log handler. Thoughts?

That way we could also re-initialize our internal logger as many times we want with different filter levels etc etc.

Reply:
Yeah, we'll have entry point that initializes both loggers based on configuration / env variables